### PR TITLE
Jekyll clip

### PR DESCRIPTION
**Pull request by hpjohn**
Create modifiable group clipping

Add a list of clipping areas to map file using new editor window (enable cl_editorcliptrigger 1)
 -These consist of tune zone, trigger id, clip x,y,w&h, disableclipping and rewind (envelopes)
Assign a trigger to groups that you want to change (in group context popup)
Add tune zones as normal

Current ver DDNet server & client load these map fine (but with clipping fixed as default)
Vanilla TW connects to server running these maps also

### CODE REVIEW

File: src/game/editor/editor.cpp [hunk 1]
```diff
@@ -4916,6 +4933,394 @@ void CEditor::RenderServerSettingsEditor(CUIRect View)
...
+
+
+	//Preview the clip areas
+	UI()->ClipEnable(&EditorRect);
+	m_Map.m_pGameGroup->MapScreen();
+	Graphics()->TextureSet(-1);
+
+	if (g_Config.m_ClEditorClipDraw >= 1) {
+		CEditorMap::CClipTrigger previewClip = m_aClipTrigger;
+		if (previewClip.disable)continue;
```
Comment (H-M-H) [inline src/game/editor/editor.cpp]: Our CI tells me there is no loop, so continue can not be used here.

File: src/game/mapitems.h [hunk 2]
```diff
@@ -412,6 +414,18 @@ struct CMapItemSound
 	int m_SoundDataSize;
 } ;
 
+struct CMapItemClips
+{
+	int zone;
```
Comment (H-M-H) [inline src/game/mapitems.h]: For members please use: `m_...`

**Comment by H-M-H** [comment]
First of all thanks for your interest into improving DDNet. Can you please elaborate a bit more on how everything works ? How exactly do you store the list of clipping areas ? When are envelopes rewound, you somehow use tune zones for that right ? How do you edit all this in editor ?

Also a general note on codingstyle: Curly braces `{}` get their own line and everything should be CamelCase.

**Comment by Ryozuki** [comment]
Source formatting guide https://www.teeworlds.com/?page=docs&wiki=nomenclature

**Comment by Learath2** [comment]
You might want to try better messages for your commits, seems like you copy pasted this to github, which is a really quick way to making mistakes :) 

Try cloning the source and using git, it's not that difficult really.

**Comment by heinrich5991** [comment]
> You might want to try better messages for your commits, seems like you copy pasted this to github, which is a really quick way to making mistakes :)
> 
> Try cloning the source and using git, it's not that difficult really.

We can help you with that if you want. Join our IRC channel or the Discord channel.

**Comment by hpjohn** [comment]
How it works: new CMapItemClips store tune zones, clipping area configuraition and trigger IDs and CMapItemTriggers pair up each group with a trigger number. On entering a new tune zone (client side) the 
CMapItemClips with the same tune zone number are all passed to groups with matching trigger number

Data is inputted in the editor with a new panel at the bottom (enable cl_editorcliptrigger 1)
data is saved in file as new items with no modification to existing mapitems, making compatibility with old servers (data is ignored) and old clients connected to new servers (nothing cahnged serverside)

@Learath2 How deatiled should commit messages be? I put a breif explaination of what each commit covered. 
Also yes, it was just copy/paste, because auto format put a million spaces into every doc (a||b) >  (a || b) etc

**Comment by Learath2** [comment]
First line of the commit message is a brief explanation in 50 characters. If you feel like you need to describe it more just leave an empty line and add a detailed explanation.

**Comment by def-** [comment]
@Learath2 What's with this change?

**Pull request closed without merging**